src/components/Presenter.js

```diff
@@ -0,0 +1,358 @@
+      }
+      this.timerChange = setTimeout(this.handleEnd, timeout);
+    }
+  }
+
+  trickyEnd() {
+    return setTimeout(() => {
+      const next = this.loadNextItem();
+      this.state.current.node.removeEventListener('transitionend', this.handleNodeAnimEnd);
+      this.setState({
+        index: next.index,
+        current: {
+          item: next.item,
+          status: ItemStatus.LOADED,
+        },
+      });
+    }, 20);
+  }
+
+  setupNode(node) {
+    const { current } = this.state;
+    if (node && !current.node && current.status === ItemStatus.LOADED) {
+      const DOMNode = ReactDOM.findDOMNode(node);
+      this.setState(
+        {
+          current: {
+            ...this.state.current,
+            node: DOMNode,
+          },
+        },
+        () => {
+          this.trickyStart();
+        }
+      );
+
+      DOMNode.addEventListener('transitionend', this.handleNodeAnimEnd, false);
+    }
+  }
+
+  handleNodeAnimEnd(e) {
+    const { current: { node, status } } = this.state;
+    if (status === ItemStatus.ENTERING || status === ItemStatus.LEAVING) {
+      const { timeout } = this.props;
+      const ns = status + 1;
+
+      this.setState(
+        {
+          current: {
+            ...this.state.current,
+            status: ns,
+          },
+        },
+        () => {
+          if (this.state.status === PresenterStatus.PLAYING) {
+            if (ns === ItemStatus.ENTERED) {
+              this.timerChange = setTimeout(this.handleEnd, timeout);
+            } else if (ns === ItemStatus.LEFT) {
+              this.trickyEnd();
+            }
+          }
+        }
+      );
+    }
+  }
+
+  getStyle(status) {
+    const { transitionDuration, transitions } = this.props;
+    switch (status) {
+      case ItemStatus.LOADED:
+        return {
+          transition: 'none',
+          ...transitions.enter,
+        };
+      case ItemStatus.LEFT:
+        return {
+          ...transitions.enter,
+          transition: 'none',
+        };
+      case ItemStatus.ENTERING:
+        return {
+          transition: transitionDuration + 'ms',
+          ...transitions.normal,
+        };
+      case ItemStatus.ENTERED:
+        return transitions.normal;
+      case ItemStatus.LEAVING:
+        return {
+          transition: transitionDuration + 'ms',
+          ...transitions.left,
+        };
+    }
+  }
+
+  render() {
+    const { current, next, status } = this.state;
+    const { transitionDuration, rightBar } = this.props;
+    if (current && current.status >= ItemStatus.LOADED) {
+      const st = this.getStyle(current.status);
+      return (
+        <div className="presenter">
+          <div className="viewer">
+            <PresenterItem
+              transitionDuration={transitionDuration}
+              stateStyle={st}
+              ref={node => this.setupNode(node)}
+              item={current.item}
+            />
+          </div>
+          <div className="bar">
```

Comment:
I am quite confident this should be a separate component

Reply:
Exactly. It needs to be promoted. However, I am not sure if I am going to change for this PR. Perhaps, some next one.